### PR TITLE
don't use @ os_only macros

### PR DESCRIPTION
**Pull request by pfitzseb**
also, `OS_NAME` doesn't exist anymore

### CODE REVIEW

File: src/BinDeps.jl [hunk 1]
```diff
@@ -412,22 +416,32 @@ module BinDeps
...
-    	@windows_only @dependent_steps begin
-    		begin
-                ChangeDirectory(s.src)
-    			@windows_only FileRule(isempty(s.config_status_dir)?"config.status":joinpath(s.config_status_dir,"config.status"),setenv(`sh -c $cmdstring`,env))
-                FileRule(s.libtarget,MakeTargets())
-                MakeTargets("install")
+    	@static if is_windows()
+            @dependent_steps begin
+        		begin
+                    ChangeDirectory(s.src)
```
Comment (nalimilan) [inline src/BinDeps.jl]: Wrong indentation here.

**Comment by nalimilan** [comment]
Thanks! Though we'll have to add compatibility macros to Compact.jl before merging this.

File: src/BinDeps.jl (hunk 1 above)
Comment (pfitzseb) [inline src/BinDeps.jl]: Yeah, whitespace was messed up in those files. Thanks for pointing that out!

File: src/dependencies.jl [hunk 2]
```diff
@@ -603,11 +602,11 @@ function check_system_handle!(ret,dep,handle)
 end
 
 # Default installation method
-if OS_NAME == :Darwin
+@static if is_apple()
```
Comment (tkelman) [inline src/dependencies.jl]: shouldn't need to be static

File: src/BinDeps.jl [hunk 3]
```diff
@@ -47,7 +47,7 @@ module BinDeps
     function download_cmd(url::AbstractString, filename::AbstractString)
         global downloadcmd
         if downloadcmd === nothing
-            for download_engine in @windows? (:powershell, :curl, :wget, :fetch) : (:curl, :wget, :fetch)
+            for download_engine in @static is_windows() ? (:powershell, :curl, :wget, :fetch) : (:curl, :wget, :fetch)
```
Comment (tkelman) [inline src/BinDeps.jl]: only need to use `@static` for ccall's

File: src/BinDeps.jl [hunk 4]
```diff
@@ -95,7 +95,7 @@ module BinDeps
         end
     end
 
-    @windows_only begin
+    @static if is_windows()
```
Comment (tkelman) [inline src/BinDeps.jl]: just `else` on the above

File: src/BinDeps.jl [hunk 5]
```diff
@@ -298,12 +298,12 @@ module BinDeps
...
     (|)(a::BuildStep,b::BuildStep) = SynchronousStepCollection()
     function (|)(a::SynchronousStepCollection,b::SynchronousStepCollection)
-    	if(a.cwd==b.cwd)
-  		append!(a.steps,b.steps)
-    	else
-    		push!(a.steps,b)
-    	end
-    	a
+        if(a.cwd==b.cwd)
+          append!(a.steps,b.steps)
```
Comment (tkelman) [inline src/BinDeps.jl]: may as well fix the spacing all the way

File: src/BinDeps.jl [hunk 6]
```diff
@@ -361,27 +361,31 @@ module BinDeps
...
+            end
+            if(!isempty(a.targets))
+                cmd = `$cmd $(a.targets)`
+            end
+            @dependent_steps ( setenv(cmd, adjust_env(a.env)), )
         end
-        @dependent_steps ( setenv(cmd, adjust_env(a.env)), )
     end
-    @windows_only lower(a::MakeTargets,collection) = @dependent_steps ( setenv(`make $(!isempty(a.dir)?"-C "*a.dir:"") $(a.targets)`, adjust_env(a.env)), )
+    @static if is_windows()
```
Comment (tkelman) [inline src/BinDeps.jl]: just `else`

File: src/BinDeps.jl [hunk 7]
```diff
@@ -412,47 +416,57 @@ module BinDeps
...
             end
         end
 
-    	@windows_only @dependent_steps begin
-    		begin
-                ChangeDirectory(s.src)
-    			@windows_only FileRule(isempty(s.config_status_dir)?"config.status":joinpath(s.config_status_dir,"config.status"),setenv(`sh -c $cmdstring`,env))
-                FileRule(s.libtarget,MakeTargets())
-                MakeTargets("install")
+        @static if is_windows()
```
Comment (tkelman) [inline src/BinDeps.jl]: `else`

**Comment by pfitzseb** [comment]
Alright, I've cleaned this up as per the comments (thanks @tkelman!) and gotten rid of a few more `@static`s -- IIUC, all usages of `@static` at the top level are unnecessary, but inside of method definitions there's a performance win because the `if` doesn't have to be evaluated at runtime, correct?
If so, I think I've used it correctly now.

**Comment by tkelman** [comment]
I just merged `@static` into Compat, it'll need to be tagged though (then that tag's version used in REQUIRE here). And I don't think we've figured out the Compat story for `is_windows` etc yet - I believe those existed in Base but not exported before?

**Comment by pfitzseb** [comment]
[Not really](https://github.com/JuliaLang/julia/commit/694413abca84b1745afe000f05a1676e899e6a43) -- there was only `is_unix`. So we need to add all those function to Compat?

**Comment by tkelman** [comment]
Yeah, you're right, my memory was faulty there. Yes these need to be implemented in Compat.

File: src/BinDeps.jl [hunk 8]
```diff
@@ -47,7 +47,7 @@ module BinDeps
     function download_cmd(url::AbstractString, filename::AbstractString)
         global downloadcmd
         if downloadcmd === nothing
-            for download_engine in @windows? (:powershell, :curl, :wget, :fetch) : (:curl, :wget, :fetch)
+            for download_engine in (is_windows() ? (:powershell, :curl, :wget, :fetch) : (:curl, :wget, :fetch))
```
Comment (pfitzseb) [inline src/BinDeps.jl]: Without the parentheses around the whole ternary this errors with 
```
ERROR: LoadError: syntax: space before "(" not allowed in "? ("
 in include_from_node1(::String) at .\loading.jl:426
 in macro expansion; at .\none:2 [inlined]
 in anonymous at .\<missing>:?
 in eval(::Module, ::Any) at .\boot.jl:225
 in process_options(::Base.JLOptions) at .\client.jl:243
 in _start() at .\client.jl:322
while loading C:\Users\...\.julia\v0.5\BinDeps\src\BinDeps.jl, in expression starting on line 2
```
but works fine if there's no space... Not quite sure what to make of it.

Comment (tkelman) [inline src/BinDeps.jl]: That sounds like it may be a parsing bug? Or depending on what gets tokenized first, `?` might be getting treated as a function.

Comment (pfitzseb) [inline src/BinDeps.jl]: The latter, I think:
```
julia> for a in is_windows() ? [1 2] : [3 4]
ERROR: syntax: space before "[" not allowed in "? ["
 in eval(::Module, ::Any) at .\boot.jl:225
 in macro expansion at .\REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at .\event.jl:46

julia> for a in is_windows() ?[1 2] : [3 4]
       println(a)
       end
ERROR: UndefVarError: ? not defined
 in macro expansion; at .\REPL[9]:1 [inlined]
 in anonymous at .\<missing>:?
 in eval(::Module, ::Any) at .\boot.jl:225
 in macro expansion at .\REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at .\event.jl:46

julia> Base.operator_precedence(:in)
6

julia> Base.operator_precedence(:?)
2
```
So this seems like correct, albeit kinda confusing on first glance.

Comment (tkelman) [inline src/BinDeps.jl]: maybe in 0.6 some sanity can be restored to `?` now that it isn't being punned on with these `@windows?` etc fake ternary macros

**Comment by TotalVerb** [comment]
:+1: This would fix building binary dependencies on OSX, which is currently broken on nightly because `is_unix` expects `:Darwin` but the deprecated `Base.OS_NAME` is `:OSX`. I think that's the cause of many spurious Travis failures on OSX nightly in downstream packages.

**Comment by andreasnoack** [comment]
Superseded by #229 and #234

**Pull request closed without merging**